### PR TITLE
Added support for `torch.Tensor.unfold`

### PR DESCRIPTION
**Pull request by riccardofelluga**
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

</details>

## What does this PR do?

This PR adds support for `torch.Tensor.unfold`

## Did you have fun?

Oh yes! 🚀

cc @apaz-cli

### CODE REVIEW

File: thunder/executors/torchex.py [hunk 1]
```diff
@@ -553,6 +554,7 @@ def _squeeze_transform(a: TensorLike, /, dim: None | int | Sequence[int] = None)
 _register_implementation(ltorch.tensor_split, tensor_split, checker=_always_executable)
 _register_implementation(ltorch.transpose, transpose, checker=_always_executable)
 _register_implementation(ltorch.unbind, unbind, checker=_always_executable)
+_register_implementation(ltorch.unfold, unfold, checker=_always_executable)
 _register_implementation(ltorch.unsqueeze, unsqueeze, checker=_always_executable)
```
Comment (nikitaved) [inline thunder/executors/torchex.py:558]: Are we not missing implementations for clang/prim symbols?

Comment (riccardofelluga) [reply thunder/executors/torchex.py:559]: Yupp sorry my bad!

Comment (nikitaved) [reply thunder/executors/torchex.py:559]: Let's test against `clang.unfold` then? Just to be safe

File: thunder/core/prims.py [hunk 2]
```diff
@@ -3077,6 +3078,27 @@ def transpose_meta(a: TensorProxy, /, permutation: tuple[int, ...]) -> TensorPro
...
+    utils.check(
+        size <= max_size, lambda: f"Maximum size for tensor at dimension {dim} is {max_size} but size is {size}"
+    )
+    utils.check(size >= 0, lambda: f"Size is {size} but must be >= 0")
+    utils.check(step > 0, lambda: f"Step is {step} but must be > 0")
+
+    shape = list(a.shape)
+    shape.append(size)
+    if dim < a.ndim:
+        shape[dim] = (shape[dim] - size) // step + 1
```
Comment (nikitaved) [inline thunder/core/prims.py:3095]: Dim is already canonicalized, so there is no need in this `if`.

File: thunder/executors/torchex.py (hunk 1 above)
Comment (riccardofelluga) [reply thunder/executors/torchex.py:559]: You mean something like this?

``` python
In [1]: import torch; import thunder

In [2]: a = torch.arange(9)

In [3]: a
Out[3]: tensor([0, 1, 2, 3, 4, 5, 6, 7, 8])

In [4]: def f(x):
   ...:     return thunder.clang.unfold(x, 0, 2, 2)
   ...: 

In [5]: jf = thunder.jit(f, only_execute_prims=True)

In [6]: jf(a)
Out[6]: 
tensor([[0, 1],
        [2, 3],
        [4, 5],
        [6, 7]])
```

Comment (nikitaved) [reply thunder/executors/torchex.py:559]: I mean the op that we test in OpInfos is probably safer to change from `ltorch.unfold` to `clang.unfold`.

Comment (riccardofelluga) [reply thunder/executors/torchex.py:559]: Ohh I see! Is there a reason why sometimes in the tests we test against ltorch instead of clang?

Comment (nikitaved) [reply thunder/executors/torchex.py:559]: Not all ops in ltroch are in clang. And we still have a gap in testing decompositions unless the backward rule is implemented...

File: thunder/tests/opinfos.py [hunk 3]
```diff
@@ -2994,6 +2994,43 @@ def unbind_sample_generator(op, device, dtype, requires_grad, **kwargs):
 shape_ops.append(unbind_opinfo)
 
 
+def unfold_sample_generator(op, device, dtype, requires_grad, **kwargs):
+    make = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+
+    cases = (
+        ((0,), 0, 0, 1),
```
Comment (nikitaved) [inline thunder/tests/opinfos.py:3001]: Let's add some tests for scalars, with different values of `size/step`. So that the line `max_size = 1 if a.ndim == 0 else a.shape[dim]` is justifiable.

**Review by nikitaved — APPROVED** [review]
LGTM! Thank you, @riccardofelluga !

**Comment by IvanYashchuk** [comment]
Codeowner approval is required to merge.
@t-vi, @carmocca

**Review by carmocca — APPROVED** [review]
Thanks!
